File: src/main/java/com/protryon/jasm/stage1/constant/CUTF8.java
package com.protryon.jasm.stage1.constant;

import java.io.DataOutputStream;
import java.io.IOException;

import com.protryon.jasm.stage1.CConstant;

public class CUTF8 extends CConstant<String> {
	public CUTF8(String value) {
		super(value, 1);
	}

	@Override
	protected void write(DataOutputStream out) throws IOException {
		byte[] utf = this.value.getBytes();
		out.writeShort(utf.length);
		out.write(utf);
	}

	public static CUTF8 assertCUTF8(CConstant CConstant) {
		if (!(CConstant instanceof CUTF8)) {
			throw new RuntimeException("Expected UTF8!");
		}
		return (CUTF8) CConstant;
	}
}

File: src/main/java/com/protryon/jasm/stage1/constant/CUTF8.kt
package com.protryon.jasm.stage1.constant

import java.io.DataOutputStream
import java.io.IOException

import com.protryon.jasm.stage1.CConstant

class CUTF8(value: String) : CConstant<String>(value, 1) {

    @Throws(IOException::class)
    override fun write(out: DataOutputStream) {
        val utf = this.value.toByteArray()
        out.writeShort(utf.size)
        out.write(utf)
    }

    companion object {

        fun assertCUTF8(CConstant: CConstant<*>): CUTF8 {
            if (CConstant !is CUTF8) {
                throw RuntimeException("Expected UTF8!")
            }
            return CConstant
        }
    }
}
